File: backend/src/lib/result.ts
/**
 * Result Pattern
 *
 * Functions return Result<T, E> instead of throwing.
 * This forces explicit error handling.
 *
 * Usage:
 *   // In service
 *   function getUser(id: string): Result<User, AppError> {
 *     const user = await db.user.findUnique({ where: { id } });
 *     if (!user) return err(Errors.USER_NOT_FOUND);
 *     return ok(user);
 *   }
 *
 *   // In controller
 *   const result = await getUser(id);
 *   if (!result.ok) {
 *     return c.json({ error: result.error.message }, result.error.httpStatus);
 *   }
 *   return c.json({ data: result.data });
 */

import type { AppError } from './errors';

// Success type
interface Success<T> {
  ok: true;
  data: T;
}

// Failure type
interface Failure<E> {
  ok: false;
  error: E;
}

// Result is either Success or Failure
export type Result<T, E = AppError> = Success<T> | Failure<E>;

// Helper to create success result
export function ok<T>(data: T): Success<T> {
  return { ok: true, data };
}

// Helper to create failure result
export function err<E>(error: E): Failure<E> {
  return { ok: false, error };
}

/**
 * Wrap an async function that might throw into a Result
 *
 * Useful for wrapping external APIs (Razorpay, Cashfree, Cloudinary, etc.)
 *
 * Usage:
 *   const result = await tryCatch(
 *     () => razorpay.subscriptions.create(options),
 *     Errors.PAYMENT_FAILED
 *   );
 */
export async function tryCatch<T>(
  fn: () => Promise<T>,
  errorOnFail: AppError
): Promise<Result<T, AppError>> {
  try {
    const data = await fn();
    return ok(data);
  } catch {
    return err(errorOnFail);
  }
}

/**
 * Unwrap a Result, throwing if it's an error
 *
 * Only use this when you KNOW it should succeed
 * (e.g., after a check, or in tests)
 */
export function unwrap<T>(result: Result<T, unknown>): T {
  if (!result.ok) {
    throw new Error('Unwrapped an error result');
  }
  return result.data;
}

File: packages/rescue_core/lib/error/result.dart
import 'api_error.dart';

/// Result Type
///
/// Functional error handling pattern - no throwing!
/// Forces explicit handling of success and failure cases.
///
/// Usage:
///   Result<User, ApiError> result = await getUser();
///   result.when(
///     ok: (user) => print('Got user: ${user.name}'),
///     err: (error) => print('Failed: ${error.message}'),
///   );

/// Sealed class representing either success (Ok) or failure (Err)
sealed class Result<T, E> {
  const Result();

  /// Pattern match on success/failure
  R when<R>({
    required R Function(T data) ok,
    required R Function(E error) err,
  });

  /// Check if result is success
  bool get isOk;

  /// Check if result is error
  bool get isErr;

  /// Get data if Ok, or null
  T? get dataOrNull;

  /// Get error if Err, or null
  E? get errorOrNull;
}

/// Success case
class Ok<T, E> extends Result<T, E> {
  final T data;
  const Ok(this.data);

  @override
  R when<R>({
    required R Function(T data) ok,
    required R Function(E error) err,
  }) =>
      ok(data);

  @override
  bool get isOk => true;

  @override
  bool get isErr => false;

  @override
  T? get dataOrNull => data;

  @override
  E? get errorOrNull => null;
}

/// Failure case
class Err<T, E> extends Result<T, E> {
  final E error;
  const Err(this.error);

  @override
  R when<R>({
    required R Function(T data) ok,
    required R Function(E error) err,
  }) =>
      err(error);

  @override
  bool get isOk => false;

  @override
  bool get isErr => true;

  @override
  T? get dataOrNull => null;

  @override
  E? get errorOrNull => error;
}


/// Type alias for API responses (Result with ApiError)
typedef ApiResult<T> = Result<T, ApiError>;
